Add nsorderedset classcluster

gcasa opened the pull request:
This change adds the NSOrderedSet and NSMutableOrderedSet class cluster to GNUstep.  All of the methods are present except those currently in Apple's beta.   Please review and let me know of any feedback you may have.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+  e = [tmpArray objectEnumerator];
+  while((o = [e nextObject]) != nil)
+    {
+      [self insertObject: o atIndex: index];
```
Whether this works or not depends on the implementation in the subclass. I have not looked yet, but would expect that insert also removed any previous instance of the same object in the ordered set.

gcasa replied:
Wrote a test.  Changing implementation.

gcasa replied:
FIxed.  I believe it works properly now.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+  id	tmp = [self objectAtIndex: index];
+
+  RETAIN(tmp);
+  [self replaceObjectAtIndex: index withObject: [self objectAtIndex: otherIndex]];
```
Again this depends on how the subclass implements this method.

gcasa replied:
I think this is okay for now.  I will leave this open.

gcasa replied:
Fixed

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+           options:(NSSortOptions)options
+   usingComparator: (NSComparator)comparator
+{
+}
```
The implementation is missing.

fredkiefer replied:
Sorry, I don't get the concept of closing a conversation, just because you don't want to address it at the moment. The comments are the way to track outstanding issues.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+
+- (void) minusOrderedSet:(NSOrderedSet *)other
+{
+  if(other != self)
```
This should be ==

gcasa replied:
Fixed

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+    }
+}
+
+- (instancetype) initWithCoder: (NSCoder *)acoder
```
Why do we need this method?

gcasa replied:
No

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+
+- (id) init
+{
+  return [self initWithObjects: 0 count: 0];
```
The first 0 should be NULL and why not have this method already in the abstract class?

gcasa replied:
Fixed

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+
+- (NSUInteger) hash
+{
+  return [self count];
```
Is this hash implementations good enough?

fredkiefer replied:
As it is good enough for the other base classes, it will do here as well.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+  return item.obj;
+}
+
+- (id) objectAtIndexedSubscript: (NSUInteger)index
```
If this is true, why not have this method already in the abstract class?

gcasa replied:
Moved

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+
+  // Reduce C if it doesn't match.  This is apparently how
+  // macOS behaves.
+  if(c > sizeof(objs))
```
I am not sure whether this will work. How is the compiler able to find out about the real size of the array?

gcasa replied:
Removed this

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+	}
+
+      item.obj = obj;
+      if(![self containsObject: obj])
```
Shouldn't we work on the array directly here?

gcasa replied:
Unfortunately there is no inline method to search for an object.  I believe the hashing method you mentioned earlier is essential for making this class as fast as possible.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+  return self;
+}
+
+- (id) initWithObject: (id)obj
```
Move this method to the abstract class

gcasa replied:
Done

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+  return[self objectAtIndex: index];
+}
+
+- (BOOL) containsObject: (id)anObject
```
This method could be moved to the abstract class.

fredkiefer replied:
Or rather not. We should make this method really fast by using a hashing structure in addition to the array. This is the main difference between this class and NSArray, so this may actually be the most important part of the class.

gcasa replied:
Okay, cool.

gcasa replied:
Since implementing the hashing structure is more involved I'm going to submit a change request for it.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+
+- (void) insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
```
Why don't we check containsObject: here?

gcasa replied:
Correct.   It should.

gcasa replied:
Done

gcasa commented on Source/NSDictionary.m:
```diff
@@ -949,6 +949,16 @@ - (NSArray*) keysSortedByValueUsingSelector: (SEL)comp
 
+- (NSArray *)keysSortedByValueUsingComparator:(NSComparator)cmptr
+{
+  return [self keysSortedByValueWithOptions:0 usingComparator:cmptr];
```
Do we have an emacs config which can help with this.  I recently did a new installation of Linux and lost my setup.

gcasa replied:
Fixed

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+- (id) mutableCopyWithZone: (NSZone*)zone
+{
+   NSMutableOrderedSet	*copy = [NSMutableOrderedSet_concrete_class allocWithZone: zone];
+   return [copy initWithOrderedSet: self copyItems: NO];
```
Because the same thing is done in NSSet.m.   I had assumed it was right since it was done there and in NSArray.m

gcasa replied:
From NSSet.m... please let me know if you still want me to change this.

/**
 * Returns a new instance containing the same objects as
 * the receiver.<br />
 * The default implementation does this by calling the
 * -initWithSet:copyItems: method on a newly created object,
 * and passing it NO to tell it just to retain the items.
 */
- (id) mutableCopyWithZone: (NSZone*)z
{
  NSMutableSet	*copy = [NSMutableSet_concrete_class allocWithZone: z];

  return [copy initWithSet: self copyItems: NO];
}

rfm replied:
IIRC the reason for this is that conventionally, copy is supposed to create an immutable copy, and does a deep copy to ensure that contents are immutable as well as the container.  A mutable copy on the other hand, is free to use the quicker/more efficient mechanism of simply retaining contents.

gcasa replied:
Okay, so this is okay.  I'm closing this comment.  Thanks, Richard.

gcasa commented on Headers/Foundation/NSArray.h:
```diff
@@ -283,7 +283,7 @@ DEFINE_BLOCK_TYPE(GSPredicateBlock, BOOL, GS_GENERIC_TYPE(ElementT),
     sortedArrayWithOptions: (NSSortOptions)options
            usingComparator: (NSComparator)comparator;
-
+  
```
Fixed.

fredkiefer replied:
But never committed :-)

gcasa replied:
I thought you were talking about the extra return. ;)

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+      self = [NSOrderedSet_concrete_class allocWithZone: NSDefaultMallocZone()];
+      return [self initWithCoder: coder];
+    }
+  else if (c == NSOrderedSet_abstract_class)
```
You're correct.  Fixed

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+    {
+      id	array;
+
+      array = [(NSKeyedUnarchiver*)coder _decodeArrayOfObjectsForKey:
```
No, I haven't.   This is pending.  I will resolve this once done.

fredkiefer replied:
I just checked the key coded archiving and your version should work.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+  return nil;
+}
+
+- (instancetype) initWithArray:(NSArray *)other copyItems:(BOOL)flag
```
Fixed

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+*/
+- (NSUInteger)_countForObject: (id)object  // required override...
+{
+  return 1;
```
The same thing is done in NSSet.m . I'm not sure why, but maybe we should review that as well.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+            }
+         else
+           {
+             if ([self _countForObject: o] != [aSet _countForObject: o])
```
Actually, this is entirely wrong.  In order to be equal the set must also be in the same sequence.

gcasa replied:
I have changed it to check if they are equal and then check if the receiver is a subset of aSet.   This should be sufficient since we are simply checking if it is a strict subset of the other and order matters.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+  if ([self count] == 0)
+    return NO;
+
+  // 0. Loop for all members in otherSet
```
This code was completely rewritten based on my tests.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+// Creating a Sorted Array
+- (NSArray *) sortedArrayUsingDescriptors:(NSArray *)sortDescriptors
+{
+  NSMutableArray *sortedArray = [GSMutableArray arrayWithArray: [self array]];
```
Fixed.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+		{
+		  rem = [self methodForSelector: remSel];
+		  /*
+		   * We need to retain the object so that when we remove the
```
This is likely okay for now.

gcasa replied:
I have added a "break;" after the removal is done.  This should stop the search once the first (and only) instance is found.

fredkiefer replied:
I do Not see the "break" and why would you cache the method that only gets called once?

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+	}
+      else
+	{
+          [other getObjects: objs];
```
It should act the same whether this is a proxy or not.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+    }
+  else
+    {
+      GS_BEGINIDBUF(objs, count);
```
Done

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+		{
+		  if(flag == YES)
+		    {		  
+		      objs[i] = [[other objectAtIndex: i] copy];
```
Done

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1781 @@
+	  {
+	    if(flag == YES)
+	      {		  
+		objs[i] = [[other objectAtIndex: i] copy];
```
This should still be objc[j]

gcasa replied:
Done

fredkiefer replied:
No, in the code I see this is still wrong.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1781 @@
+// Key value coding support
+- (void) setValue: (id)value forKey: (NSString*)key
+{
+}
```
Here the implementation is missing

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+   NSUInteger count = [self count];
+  if (count == 0)
+    return nil;
+  return [self objectAtIndex: 0];
```
Done

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+
+- (NSEnumerator *) objectEnumerator
+{
+  return nil;
```
They do now.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+{
+  id	o = nil, e = nil;
+
+  // -1. If this set is empty, this method should return NO.
```
As did I.  CS and Math and, Yes, the empty set should always be a subset.  I was thinking of that this morning.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,419 @@
+
+- (void) removeObjectAtIndex: (NSUInteger)index
+{
+  GSIArrayRemoveItemAtIndex(&array, index);
```
Shouldn't this change the _version as well?

gcasa replied:
Done

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,419 @@
+  GSIArrayItem item;
+  if([self containsObject: object] == NO)
+    {
+      item.obj = object;
```
You should write a test where you first add an object to an NSMutableOrderedSet and then insert it with a different index. Will the index of the object changes or not?

gcasa replied:
Will do

gcasa replied:
On macOS it doesn't change and doesn't insert the new object.

gcasa replied:
Works properly on GNUstep

fredkiefer replied:
It would be good to have this as a specific test too.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+  for(i = 0; i < count; i++)
+    {
+      o = objects[i];
+      [self insertObject: o atIndex: aRange.location]; 
```
I believe you’re right.  I didn’t notice this.

gcasa replied:
Fixed

fredkiefer replied:
I do not see any difference in the code. Did you forget to commit some bits?

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+
+- (BOOL)containsObject:(id)anObject // required override
+{
+  [self subclassResponsibility: _cmd];
```
Done

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+                              options:(NSEnumerationOptions)opts
+                          passingTest:(GSPredicateBlock)predicate
+{
+  return [[self objectsAtIndexes: indexSet]
```
I will write a test on macOS and compare output when the test is moved over to GNUstep

gcasa replied:
These are implemented identically on NSArray.

fredkiefer replied:
If it is the same there, it could still be wrong in both places. You wanted to write a test, what did that show?

gcasa replied:
I haven't written it yet.  I will need to create a blocks.m and build with clang.   I also need to build a test on macOS to verify.   Doing that now.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+                              options:(NSEnumerationOptions)opts
+                          passingTest:(GSPredicateBlock)predicate
+{
+  return [[self objectsAtIndexes: indexSet]
```
I will write a test on macOS and compare output when the test is moved over to GNUstep

gcasa replied:
Implemented identically on NSArray.

fredkiefer replied:
We really need these tests.

gcasa commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
```
I'm not sure why.  I looks alright to me.   Can you be more specific?

gcasa replied:
I will write a test against this to see if I'm producing the right results.

gcasa replied:
I've re-implemented this.method.  I have written a test to verify it.

fredkiefer commented on Headers/Foundation/NSOrderedSet.h:
```diff
@@ -0,0 +1,212 @@
+- (GS_GENERIC_TYPE(ElementT)) valueForKey: (NSString*)key; 
+
+// Key-Value Observing Support
+/*
```
We do not need to declare these methods here but should we implement them? Apple documents that they raise an exception. Is this really true? It should be easy to implement them similar to NSArray. Why doesn't Apple want them to be there?

gcasa replied:
For now I'm going to raise an exception as recommended in the docs, but I believe we should revisit this going forward.  It doesn't really make any sense why they don't want this to be observable.

gcasa replied:
Added

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,399 @@
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)self;
```
This method is still missing an implementation.

fredkiefer replied:
Or rather remove it an reuse the implementation from the abstract class.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,399 @@
+  return [self count];
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
```
Richard just changed the behaviour of this method in the main branch. You may need to have a look at that change.

rfm replied:
I can look at that.  I'm trying to make the method more automated (so people mostly don't need to override it in subclasses), and while the current code in the main branch works I'm not yet finished with it (I'm wondering whether to use malloc_usable_size() on non-object pointer ivars in the default code).

gcasa replied:
Marking this for later resolution.  I will add an issue and assign rfm.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,399 @@
+                                   objects: (id*)stackbuf
+                                     count: (NSUInteger)len
+{
+  //state->mutationsPtr = (unsigned long *)&_version;
```
Here the (currently missing) implementation in GSOrderedSet should work.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,399 @@
+}
+@end
+
+@interface	NSGOrderedSet : NSOrderedSet
```
I googled for these two classes but could not find any reference to them. In NSArray things are different, there Richard needs to handle the old implementation of NSArray that made its way into a few NIB files that we still want to load.

gcasa replied:
I had put them there because they were in NSArray... I wasn't sure what the use for them was.

gcasa replied:
Removed

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+- (void) setValue: (id)value forKey: (NSString*)key
+{
+  NSUInteger    i;
+  NSUInteger	count = [self count];
```
Why didn't you use an enumerator as in the method below? I always prefer to express the same idea in a similar way in different places.

gcasa replied:
Done.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+      int	p = state->state;
+      int	i;
+
+      for (i = 0; i < count; i++, p++)
```
You could actually use [self getObjects: stackbuf range: NSMakeRange(state->state, count)]; here and benefit from a better implementation of that method in a subclass.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,399 @@
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+      privateCountOfSel = @selector(_countForObject:);
```
I think this no longer gets used anywhere. You may now even drop the whole method _countForObject:.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+
+- (BOOL) intersectsSet: (NSSet *)otherSet
+{
+  id	o = nil, e = nil;
```
e should not be id but NSEnumerator instead.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+    id	o = nil, e = nil;
+
+  // -1. If this set is empty, this method should return NO.
+  if ([self count] == 0)
```
Shouldn't the implementation here be similar to the one above?

fredkiefer replied:
Why did you mark this as resolved as the code still is clearly wrong and different from the ordered case? If self is an empty set it should be a subset and the rest of the code checks for super set not subset.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+
+  // Find the first object's index in otherset, if not found
+  // it's not a subset...
+  f = [self firstObject];
```
As I understand the logic here, the subset needs to have the same objects as the super set and also in the same order but the super set may have a few extra objects in between. This is easy to implement with two enumerators. The one for self gets advanced when ever the current two objects are the same. The other enumerator gets increased for each step. If self get exhausted we return YES.  If the other is exhausted first we return NO.

gcasa replied:
Implemented as you suggest.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1814 @@
+    }
+
+  // Remove the originals...  
+  for(i = 0; i < count; i++)
```
This bit here is definitely wrong. But the rest also needs some testing with the new objects being inserted before and after the old ones.

triplef commented on Headers/Foundation/NSOrderedSet.h:
```diff
@@ -0,0 +1,212 @@
+- (instancetype) initWithCoder: (NSCoder *)coder;
+@end
+
+#endif
```
This line should probably have a comment `/* _NSOrderedSet_h_GNUSTEP_BASE_INCLUDE */` like other public headers.

Also the file is missing the closing bracket for `extern "C"` opened at the top, i.e.:

```c
#if	defined(__cplusplus)
}
#endif
```

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1912 @@
+
+- (instancetype) init
+{
+  self = [super init];
```
I think we should use the subclass implementation here and call [self initWithObjects: NULL count: 0]

gcasa replied:
done.

gcasa replied:
done.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1912 @@
+- (instancetype) initWithObjects:(const id [])objects // required override.
+                           count:(NSUInteger)count
+{
+  self = [self init];
```
Shouldn't we just raise an exception here?  [self subclassResponsibility: _cmd];

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,381 @@
+
+- (void) insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
```
The check for nil is missing here.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,381 @@
+		   withObject: (id)obj
+{
+  [self _insertObject: obj atIndex: index];
+  [self removeObjectAtIndex: index];
```
Shouldn't you remove the old object first? This looks like you remove the newly added object. Did you write a test for this case?

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,381 @@
+- (void) removeObjectAtIndex: (NSUInteger)index
+{
+  _version++;
+  GSIArrayRemoveItemAtIndex(&array, index);
```
I don't quite understand why we need to retain the object on insert, but don't release it on remove? And isn't GSIArray able to handle both itself?

rfm replied:
I agree ... Also, in the source I looked at, GSIArray is configured with macros to do retain/release, but then configured with settings to not use them.

gcasa replied:
I see what I did... I am changing the config so it does retain.  I will fix all of the issues elsewhere.

rfm commented:
I have to apologise for taking so very long before I had a chance to look at this issue at all.  even now I've really only had a quick look, concentrating on the concrete implementation.  I also had a quick skim of the Apple documentation.
One paragraph from the intro of the documentation struck me:

You can use ordered sets as an alternative to arrays when the order of elements is important and performance in testing whether an object is contained in the set is a consideration—testing for membership of an array is slower than testing for membership of a set.

This suggests to me that the current concrete class implementation (based on GSIArray) may be  a problematic approach:  perhaps the representation should fundamentally be a hash-map to rapid testing of membership, rather than an array for rapid enumeration and access by index (or both data structures could be used to hold the same objects).

If we assume that it's acceptable for access by index to be really slow for very large collections, it would be possible to implement this based on an extended version of GSIMap (or some other hash table implementation), since this contains a linked list of nodes (and of course the linked list would be used to provide ordering and access by index).  The modification required would be having an alternative version of the node addition code to maintain the linked list ordering, as well as a version of the code to resize the GSIMap (so it maintains the list ordering).

Alternatively, (simplest code approach giving high performance for both forms of access, but higher memory footprint), I'd use both a GSIMap and a GSIArray and store objects in both.  In that case it would make sense to configure one data structure to manage retain/release and the other data structure to refrain from doing retain/release.

fredkiefer commented:
I fully agree with you Richard. The current implementation does not fulfil the part documentation you cited. For me the suggestion you made in the last paragraph makes perfectly sense. But before going that way I would like to have a minimal and fully correct implementation. Optimisation should come after correctness and in my view the code is almost there.

If we get the retain policy sorted out and also add a bit of error handling, then it should be easy to just add a GSIMap for the fast lookup. Doing it first might complicate the code.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1911 @@
+@end
+
+@class	GSMutableOrderedSet;
+@interface GSMutableOrderedSet : NSObject	// Help the compiler
```
As this method is implemented on NSMutableOrderedSet it should be defined so.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1911 @@
+@end
+
+@interface NSMutableOrderedSet (Private)
+- (void) _insertObject: (id)object atIndex: (NSUInteger)index;
```
This method is only implemented on GSMutableOrderedSet and if we use it here, this creates an undocumented  requirement on subclasses. We should try to come up with an implementation without this method.

gcasa commented:
I agree with both of you.  that part of the documentation has been on my mind throughout the implementation.  Fred mentioned improving access speed early on.  I would like to have a complete and functional implementation before we focus on performance (as I believe both of you said).

fredkiefer commented on Headers/Foundation/NSOrderedSet.h:
```diff
@@ -0,0 +1,209 @@
+@class NSString;
+@class NSPredicate;
+
+@interface GS_GENERIC_CLASS(NSOrderedSet, __covariant ElementT) : NSObject <NSCoding,
```
Should we put this class into something like this?

#if OS_API_VERSION(MAC_OS_X_VERSION_10_7,GS_API_LATEST)
#endif

Other parts of base seem to use this.

gcasa replied:
Added

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,358 @@
+      if(![self containsObject: obj])
+	{
+	  GSIArrayAddItem(&array, item);
+	  RETAIN(obj);
```
If the GSIArray already retains the object, we don't need to do it here (and in all the GSMutableOrderedSet methods).

gcasa replied:
Seems to crash unless I do.  I have GSIArray configured to retain, but I'm not sure why it's losing it here.

fredkiefer replied:
This bit is important. We should not just ignore it. I have some time today and will try to look into the issue.

gcasa replied:
Please do... I want to close this out and get it merged.

gcasa replied:
It looks like you addressed this.   I would very much like to merge this at this point.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,358 @@
+  GSIArrayItem item;
+  if(object == nil)
+    {
+      NSWarnLog(@"Attempt to insert nil object");
```
Why does this give a warning log, while addObject: raises an exception?

gcasa replied:
This has been changed.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1903 @@
+// NSCoding
+- (instancetype) initWithCoder: (NSCoder *)coder
+{
+  /*
```
Could we remove this unused code?

gcasa replied:
Done.

fredkiefer replied:
Sorry, but the commented out code is still there.

gcasa replied:
GOne now... I'm assuming you removed it.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1903 @@
+  self = [self initWithObjects: objs count: 1];
+  if(self == nil)
+    {
+      NSLog(@"Problem initializing with one element");
```
Do we really want this log message? In all other cases we just hand on the result.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1903 @@
+
+- (instancetype) init
+{
+  [self subclassResponsibility: _cmd];
```
Shouldn't we use 

return [self initWithObjects: NULL count: 0];

here?

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+  self = [self initWithObjects: objs count: 1];
+  if(self == nil)
+    {
+      NSLog(@"Problem initializing with one element");
```
This specific implementation could be move to the super class.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,407 @@
+}
+
+- (void) addObject: (id)anObject
+{
```
Why the retain?

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1763 @@
+      IMP	remLast = [self methodForSelector: rlSel];
+
+      while (c--)
+	{
```
This looks like you are sorting the indices array, but it is already sorted, which make the whole method superfluouses.

gcasa replied:
This is how this seems to be done in NSArray for whatever reason.

fredkiefer replied:
Ah, here it is! The NSArray method is a (deprecated) public method. There the indexes could come in an arbitrary order. Here this is not the case.

gcasa commented:
I would very much like to close this out.  I have written bugs for the outstanding issues, but I believe that the major issues have been addressed.  Are there any major objections to merging this?

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1894 @@
+             maxCount: count
+         inIndexRange: NULL];
+
+  [self _removeObjectsFromIndices: indexArray
```
I am sure I already wrote a comment that you don't have to sort the indexes here but somehow this got lost.

gcasa replied:
I closed it because this method is actually taken from NSArray.  I'm curious though, how do we know the indexes are sorted?

gcasa replied:
I mean.. we DO in that particular method, but those indexes could come from another source.

gcasa replied:
Simplified this method as discussed.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1894 @@
+    }
+
+  // Remove all objects in this set...
+  [self removeAllObjects];
```
We should try to avoid removeAllObjects, as this will be very slow, when only a few objects change. Why not call removeObjectsAtIndexes and then loop over the indexes and insert the new objects one after another?
Or use the method replaceObjectAtIndex: withObject:

gcasa replied:
Initially I did it this way to ensure that the indexes didn't change.

gcasa replied:
Corrected

fredkiefer replied:
You changed the other place where I made this comment, but not this one.

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1894 @@
+      objs[index] = objs[otherIndex];
+      objs[otherIndex] = tmp;
+      
+      [self removeAllObjects];
```
Again, this is very inefficient. We could just remove the object at the higher index first, then the one at the lower and insert them in the opposite order.

gcasa replied:
Implemented as you suggested.

fredkiefer commented on Source/GSOrderedSet.m:
```diff
@@ -0,0 +1,332 @@
+#import "GSDispatch.h"
+#import "GSSorting.h"
+
+#define	GSI_ARRAY_TYPE	NSRange
```
OK, here is your problem with GSIArray: Whenever GSI_ARRAY_NO_RETAIN is defined, no matter to which value, there won't be any retain code called. What you need to do is remove all the following six lines, just leave the one with
#define GSI_ARRAY_TYPES       GSUNION_OBJ

Next you remove the retain calls from the code and as I tested locally your tests all still work.

fredkiefer commented:
I made a couple of clean up commits. Now most of the outstanding issues should be resolved. There is at least one missing method and two where I think the result is incorrect but you could just write bug reports for these as you suggested. And the fast enumerator code for the mutable class does not take the version field into account, this really needs fixing before going productive.

What I still would like to see here is to add a GSIMap to GSOrderedSet and use this for fast lookup. Richard could surely explain how to set up this map in a way it does not retain the elements and works just as a hash set. With that change in place this pull request would be ready to be merged.

fredkiefer commented:
Apps, looks like I broke the build. In the Travis log we have this for gcc:

Building in base/NSOrderedSet
Making all for test_tool basic...
 Compiling file basic.m ...
 Linking test_tool basic ...

Testing basic.m...
Running base/NSOrderedSet/basic.m...
Start set:       basic.m:119 ... NSOrderedSet base
Passed test:       basic.m:130 ... can create an empty ordered set
Passed test:       basic.m:135 ... can create an ordered set with one element
Passed test:       basic.m:141 ... can create an ordered set with multi element
Passed test:       basic.m:147 ... cannot create an ordered set with multiple like elements
/home/travis/staging/bin/gnustep-tests: line 309: 27621 Segmentation fault      (core dumped) $RUN_CMD
Failed file:     basic.m aborted without running all tests!

But running the tests locally with gcc worked for me. And looking at the specific test, I would expect this to fail on many machines. The test is the case where we provide an array with one element but give a count of two. How is this supposed to work?

rfm commented:
For a GSIMap configured to act as a hash table for objects with no retain/release done, you need to configure it as having no value and having the key retain/release operations do nothing.  Look at GSSet.m for an example of using this sort of setup (identical except that in GSSet.m the map does retain/release).

#define GSI_MAP_HAS_VALUE       0
#define GSI_MAP_KTYPES          GSUNION_OBJ
#define GSI_MAP_RETAIN_KEY(M, X)
#define GSI_MAP_RELEASE_KEY(M, X)
#include "GNUstepBase/GSIMap.h"

fredkiefer commented:
Thank you Richard! If Greg doesn't beat me on that, I will try to Implement this tomorrow.

As for the test issue, I just copied the code over to my Mac and there ist gives a bad access. Looks like the test code is plain wrong. I will remove this specific test. I am now also thinking about testing the whole test code on Cocoa. Any idea on how to do that efficiently?

fredkiefer commented on Source/NSOrderedSet.m:
```diff
@@ -0,0 +1,1835 @@
+  [self subclassResponsibility: _cmd];
+}
+
+- (void) _removeObjectsFromIndices: (NSUInteger*)indices
```
Why did you reintroduce this method? It is no longer used anywhere in this class and if there isn't a demonstrable case that it gets used somewhere else in public code there is no need for it.

fredkiefer replied:
After my comment you introduced the usage of this method. This really is frustrating. I wanted to help you get this polished and out to the world. I rather wait until you are through with your days work so we don't overlap.

fredkiefer commented:
Richard,

could you please have a critical look at the code again? For me most off the remaining issues I listed two days ago have now been resolved. There are still the two cases where I think that the implementation, which was copied over from NSArray, is wrong. Greg is going to write a test for this, but this seems to take some time. My approach here would be to delegate in these methods to NSArray and open a bug report for that class if the test fails :-)

gcasa commented:
I've added a test for the "indexOfObjectPassingTest:" method which uses the method in question.  It works properly.  I believe we can go ahead and merge this now.

gcasa commented:
Sorry it took some time I was working on another project for a client and also had some issues getting clang installed and building GNUstep (my own fault, I apparently made a stupid change to one of my scripts).

gcasa commented:
To delegate to NSArray I would need to convert, is that the best approach?

fredkiefer commented:
You do understand that the methods in question where
- (NSUInteger) indexOfObjectAtIndexes:(NSIndexSet *)indexSet
	                              options:(NSEnumerationOptions)opts
	                          passingTest:(GSPredicateBlock)predicate;

and 

- (NSIndexSet *) indexesOfObjectsAtIndexes:(NSIndexSet *)indexSet
	                              options:(NSEnumerationOptions)opts
	                          passingTest:(GSPredicateBlock)predicate;

whereas you provided a test for 

- (NSIndexSet *)indexesOfObjectsPassingTest:(GSPredicateBlock)predicate;

The issue with the first two methods is that the index(es) that will get returned will be in the domain of the passed in index set. This most likely isn't the intended behaviour. I will rewrite these methods to delegate to NSArray.

gcasa commented:
Both of the methods I wrote tests for use the methods in question

On Mon, Jul 1, 2019 at 13:47 Fred Kiefer <notifications@github.com> wrote:

> You do understand that the methods in question where
>
>    - (NSUInteger) indexOfObjectAtIndexes:(NSIndexSet *)indexSet
>    options:(NSEnumerationOptions)opts
>    passingTest:(GSPredicateBlock)predicate;
>
> and
>
>    - (NSIndexSet *) indexesOfObjectsAtIndexes:(NSIndexSet *)indexSet
>    options:(NSEnumerationOptions)opts
>    passingTest:(GSPredicateBlock)predicate;
>
> whereas you provided a test for
>
>    - (NSIndexSet
>    *)indexesOfObjectsPassingTest:(GSPredicateBlock)predicate;
>
> The issue with the first two methods is that the index(es) that will get
> returned will be in the domain of the passed in index set. This most likely
> isn't the intended behaviour. I will rewrite these methods to delegate to
> NSArray.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gnustep/libs-base/pull/46?email_source=notifications&email_token=AAAG2J44SPJ4SQNCYAI2XP3P5I7JLA5CNFSM4HWJG4NKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY63OEI#issuecomment-507361041>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAG2JYRVPZAUOEWUVQIWTLP5I7JLANCNFSM4HWJG4NA>
> .
>
-- 
Gregory Casamento
GNUstep Lead Developer / OLC, Principal Consultant
http://www.gnustep.org - http://heronsperch.blogspot.com
http://ind.ie/phoenix/

gcasa commented:
I’ll weite more tests for the methods you mention.

On Mon, Jul 1, 2019 at 14:38 Gregory Casamento <greg.casamento@gmail.com>
wrote:

> Both of the methods I wrote tests for use the methods in question
>
> On Mon, Jul 1, 2019 at 13:47 Fred Kiefer <notifications@github.com> wrote:
>
>> You do understand that the methods in question where
>>
>>    - (NSUInteger) indexOfObjectAtIndexes:(NSIndexSet *)indexSet
>>    options:(NSEnumerationOptions)opts
>>    passingTest:(GSPredicateBlock)predicate;
>>
>> and
>>
>>    - (NSIndexSet *) indexesOfObjectsAtIndexes:(NSIndexSet *)indexSet
>>    options:(NSEnumerationOptions)opts
>>    passingTest:(GSPredicateBlock)predicate;
>>
>> whereas you provided a test for
>>
>>    - (NSIndexSet
>>    *)indexesOfObjectsPassingTest:(GSPredicateBlock)predicate;
>>
>> The issue with the first two methods is that the index(es) that will get
>> returned will be in the domain of the passed in index set. This most likely
>> isn't the intended behaviour. I will rewrite these methods to delegate to
>> NSArray.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/gnustep/libs-base/pull/46?email_source=notifications&email_token=AAAG2J44SPJ4SQNCYAI2XP3P5I7JLA5CNFSM4HWJG4NKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODY63OEI#issuecomment-507361041>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAAG2JYRVPZAUOEWUVQIWTLP5I7JLANCNFSM4HWJG4NA>
>> .
>>
> --
> Gregory Casamento
> GNUstep Lead Developer / OLC, Principal Consultant
> http://www.gnustep.org - http://heronsperch.blogspot.com
> http://ind.ie/phoenix/
>
-- 
Gregory Casamento
GNUstep Lead Developer / OLC, Principal Consultant
http://www.gnustep.org - http://heronsperch.blogspot.com
http://ind.ie/phoenix/

gcasa commented:
As you suspected...

base/NSOrderedSet/blocks.m:
Failed test:       blocks.m:66 ... indexOfObjectAtIndexes:... Returns correct index
--- Running tests in base/NSPointerArray ---


This test succeeds on macOS Cocoa.

gcasa commented:
Corrected functionality for the two methods in question.  It was as you warned.... they were not pulling the right indexes into the result.  I updated both methods in NSArray to act properly as they do on Mac OS X,   The matching macOS tests are here: https://github.com/gcasa/NSOrderedSetTests

gcasa commented:
Thank you both for all of your input.  Merging.
